Add support for the new SQLite driver

JanJakes opened the pull request:
This PR adds support for the [new SQLite driver](https://github.com/WordPress/sqlite-database-integration/releases/tag/v2.2.0).

When the new SQLite driver is available and enabled, it will execute all the MySQL queries using the new driver. Otherwise, the legacy SQLite driver will be used.

JanJakes committed:
Add support for the new SQLite driver

JanJakes committed:
Use MySQL SHOW TABLES insteand of SQLite schema tables

JanJakes committed:
Fix linter errors

wojtekn commented on src/Export.php:
```diff
@@ -144,8 +152,12 @@ protected function write_insert_statements( $handle, $table_name ) {
+			: $table_name;
+
+		$create = $this->driver->query( 'SHOW CREATE TABLE ' . $table_name );
+		return $create[0]->{'Create Table'} . ";\n";
```
Should we `trim()` the produced query before we add a semicolon? In my tests, it ends up having a double semicolon at the end of each `CREATE TABLE` query.

JanJakes replied:
@wojtekn Oh, right, that happens when using the old translator, doesn't it? MySQL doesn't end the output of `SHOW CREATE TABLE` with a semicolon and neither does the new driver, but I guess the old driver adds it, and then we end up with a double semicolon.

wojtekn replied:
I double-checked it, and you're right - it now happens with the old driver.

JanJakes replied:
Fixed in https://github.com/Automattic/wp-cli-sqlite-command/pull/10/commits/8ee70fb63d5de90026848827576f6e54f2986060.

wojtekn commented:
@JanJakes, thanks for adding support for AST to the SQLite command. I used those steps to test it in Studio:

1. Clone the PR locally
2. Run `composer install --no-dev --optimize-autoloader --ignore-platform-reqs` in the `wp-cli-sqlite-command` directory
3. Replace `wp-files/sqlite-command` in the Studio code with a symlink to your local `wp-cli-sqlite-command` directory
4. Apply 3658f-pb patch to the Studio repo
5. Comment out the line that updates the SQLite command:
```
//await updateLatestSQLiteCommandVersion();
```
6. Run `npm start`
7. Click on the Import / Export tab
8. Click on the "Export database"
9. Observe .sql file is created
10. Open the .sql file and validate that the dump is correct

I couldn't make it fully work yet. I needed to make a few changes, see inline comments.

wojtekn commented on src/SQLiteDatabaseIntegrationLoader.php:
```diff
@@ -85,10 +85,31 @@ public static function load_plugin() {
+			require_once $plugin_directory . '/wp-includes/sqlite-ast/class-wp-sqlite-driver-exception.php';
+			require_once $plugin_directory . '/wp-includes/sqlite-ast/class-wp-sqlite-information-schema-builder.php';
+			require_once $plugin_directory . '/wp-includes/sqlite-ast/class-wp-sqlite-information-schema-exception.php';
+			require_once $plugin_directory . '/wp-includes/sqlite-ast/class-wp-sqlite-information-schema-reconstructor.php';
```
```suggestion
			require_once $plugin_directory . '/wp-includes/sqlite-ast/class-wp-sqlite-information-schema-reconstructor.php';
require_once $plugin_directory . '/wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php';
require_once $plugin_directory . '/version.php';
```

wojtekn replied:
I needed to load those two files, as it complained about missing functions and missing constants.

JanJakes replied:
@wojtekn Ah, good catch! I was testing it within a context of a WP site that was using the SQLite driver as well, so it likely had everything already required. I will check if there are any other files that may be needed.

JanJakes replied:
Fixed in https://github.com/Automattic/wp-cli-sqlite-command/pull/10/commits/b125d21678f98ea3a4cdf7b255ef1c8293eefecf.

wojtekn commented on src/SQLiteDriverFactory.php:
```diff
@@ -0,0 +1,30 @@
+	 */
+	public static function create_driver() {
+		$new_driver_supported = class_exists( WP_SQLite_Driver::class );
+		$new_driver_enabled   = $new_driver_supported && defined( 'WP_SQLITE_AST_DRIVER' ) && WP_SQLITE_AST_DRIVER;
```
In my tests, `DB_NAME` and `WP_SQLITE_AST_DRIVER` constants are undefined. I needed to hardcode the first one as `database_name_here` and `$new_driver_enabled` as `true` for testing purposes.

JanJakes replied:
@wojtekn The "correct" value of the database name will be needed with `wp sqlite import` because `CREATE TABLE` statements need to write it in the information schema. Are we in a context where importing the `wp-config.php` of a given site would be conceivable?

wojtekn replied:
The command runs on `before_wp_load`, so `wp-config.php` is not loaded. See https://github.com/Automattic/wp-cli-sqlite-command/pull/7 for some context.

Perhaps we could load it manually, similar to how WP CLI runner already does?

```
eval( WP_CLI::get_runner()->get_wp_config_code() );
```

Or alternatively, to limit the amount of loaded code, what if we used the WP CLI parameter instead of the `WP_SQLITE_AST_DRIVER` constant to enable the new driver? Like `--driver=[ast|legacy]`?

JanJakes replied:
> Or alternatively, to limit the amount of loaded code, what if we used the WP CLI parameter instead of the WP_SQLITE_AST_DRIVER constant to enable the new driver? Like --driver=[ast|legacy]?

@wojtekn This would work for the AST/legacy and would be temporary anyway. But what about the `DB_NAME`? 🤔

There's this edge-case scenario:
1. Someone imports a DB (we use some DB name during that process).
2. Then a plugin code runs `SELECT * FROM information_schema.tables WHERE table_schema = 'my-db-name' AND ...`.

In point 2, we'd like the DB name to match the `DB_NAME` that was set during the import.

By the way, is Playground booted anywhere before the import command is run? Because we know that site exports without `DB_NAME` exist, and the fallback to inject their value in `wp-config.php` is done in Playground at the moment.

JanJakes replied:
@wojtekn Looking at [`wp db import`](https://github.com/wp-cli/db-command/blob/main/src/DB_Command.php), it seems to me they're using `after_wp_config_load` on the class level, and no other annotation for the `import` command. A couple of commands use `after_wp_load`.

Also explained here: https://developer.wordpress.org/cli/commands/db/import/

> This command runs on the after_wp_config_load hook, after wp-config.php has been loaded into scope. Runs SQL queries using DB_HOST, DB_NAME, DB_USER and DB_PASSWORD database credentials specified in wp-config.php. This does not create database by itself and only performs whatever tasks are defined in the SQL.

Maybe we could use `after_wp_config_load` for the `import` command?

wojtekn replied:
@JanJakes, it makes sense to try with the `after_wp_config_load` for the `import` command, especially since it would be consistent with `wp db import`.

JanJakes replied:
@wojtekn I [pushed that change](https://github.com/Automattic/wp-cli-sqlite-command/pull/10/commits/f2ea941e1814e27f3c744ffe72f9d55334ee6240), let me know if it works for you.

wojtekn replied:
@JanJakes I realized I was testing Export above, so the change doesn't fix the behavior. When I set `after_wp_config_load` for `export` command, it reads DB_NAME from `wp-config.php` correctly.
 
However, it still doesn't read `WP_SQLITE_AST_DRIVER` that is set in Studio through `defineWpConfigConsts`.

wojtekn replied:
`WP_SQLITE_AST_DRIVER` constant set in Studio through `defineWpConfigConsts` enables the driver for the Studio site running in a browser.

It doesn't work for `sqlite export` as it is executed through a non-Playground process. Would it make sense to use  `--driver=[ast|legacy]` for that part?

JanJakes replied:
@wojtekn Great! Let's use `after_wp_config_load` for all the commands, then? As for AST, yes, let's make it configurable on the CLI 👍

wojtekn commented on src/Tables.php:
```diff
@@ -31,15 +26,19 @@ protected function get_pdo() {
+		// With the legacy driver, we need to exclude system tables
+		// and make sure the table names are alphabetically sorted.
+		if ( $this->driver instanceof WP_SQLite_Translator ) {
+			$tables_to_exclude = array( '_mysql_data_types_cache', 'sqlite_sequence' );
```
Should we also exclude `_wp_sqlite_mysql_information_schema_tables` , `_wp_sqlite_mysql_information_schema_columns`, `_wp_sqlite_mysql_information_schema_statistics`, and `_wp_sqlite_mysql_information_schema_table_constraints`?

JanJakes replied:
@wojtekn It would only make sense if some sort of downgrade could occur — that is, if `wp sqlite tables` would be run with the old driver against a SQLite database that was created with the new driver.

That said, it's easy to be on the safe side, so I can add that change.

JanJakes replied:
Fixed in https://github.com/Automattic/wp-cli-sqlite-command/pull/10/commits/0ac5622b4419d2689d593e79903c11daa1f74cc2.

JanJakes committed:
Ensure CREATE TABLE statements are terminated with a single semicolon

JanJakes committed:
Explicitly exclude internal tables used by the new driver

JanJakes committed:
Add missing file imports

JanJakes commented:
@wojtekn I pushed some changes to fix all the issues, except for `DB_NAME` — [I'm wondering if there's a way to use the correct database name in that case](https://github.com/Automattic/wp-cli-sqlite-command/pull/10#discussion_r2169113435).

JanJakes committed:
Execute "wp sqlite import" on "after_wp_config_load"

wojtekn committed:
Move export command to after wp config load to access db name

wojtekn committed:
Add param allowing to enable AST driver

wojtekn committed:
Remove incorrect check for new_driver_supported

wojtekn committed:
Fix empty space formatting

wojtekn committed:
Disable sniff that checks for constant prefix

wojtekn commented:
I pushed a few more changes:
- `--enable-ast-driver` param to enable the AST driver for CLI calls
- simplified `$new_driver_enabled` check to fix loading order issues
- moved commands to `after_wp_config_load` event

wojtekn committed:
Move commands back to before_wp_load event

wojtekn commented:
I reverted `after_wp_config_load` event change as it's no longer needed after https://github.com/WordPress/sqlite-database-integration/pull/203 was merged.

JanJakes committed:
Use "after_wp_config_load" for all SQLite commands

This ensures the DB_NAME constant from "wp-config.php" is used,
and aligns these commands with similar "wp db" commands.

With https://github.com/WordPress/sqlite-database-integration/pull/213,
the correctness of the database name will also be verified.

sejas reviewed (approved):
It works great with sqlite database integration 2.2.3 and this Studio PR https://github.com/Automattic/studio/pull/1506
